pkg/promtail/targets/filetarget.go

```diff
@@ -198,12 +171,103 @@ func (t *FileTarget) run() {
 			}
 		case err := <-t.watcher.Errors:
 			level.Error(t.logger).Log("msg", "error from fswatch", "error", err)
+		case <-ticker.C:
+			err := t.sync()
+			if err != nil {
+				level.Error(t.logger).Log("msg", "error running sync function", "error", err)
+			}
 		case <-t.quit:
 			return
 		}
 	}
 }
 
+func (t *FileTarget) sync() error {
```

Comment:
This function is a bit of a beast, and might benefit from breaking up.

Also, so set arithmetic helpers function (like subtraction, a - b) might help cleanup the logic:

```
dirs, files := glob()

toStartWatching := sub(t.watches, dirs)
startWatching(toStartWatching)

toStopWatching := sub(dirs, t.watches)
stopWatching(toStopWatching)

toStartTailing := sub(t.tailers, files)
startTailing(toStartTailing)

toStopTailing := sub(files, t.tailers)
stopTailing(toStopTailing)
```

Might make it more obvious and easier to follow.  Just a thought.